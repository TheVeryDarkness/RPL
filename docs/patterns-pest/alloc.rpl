pattern alloc

patt {
    p[$T: type] = 
        alloc_cast_write[$T = $T] 
        - alloc_check_cast_write[$T = $T] 
        - alloc_cast_check_write[$T = $T] 
        - alloc_cast_check_as_write[$T = $T]
}

util {
    alloc_cast_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_check_cast_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _2
            let $const_ptr_1: *const u8 = copy $ptr_1 as *const u8 (PtrToPtr); // _19
            let $addr_1: usize = copy $const_ptr_1 as usize (Transmute); // _20
            // It's weird that `$ptr_2` can only be declared before `switchInt`
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            let $ptr_2: *mut $T = copy $ptr_1 as *mut $T (PtrToPtr); // _4
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_cast_check_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            let $const_ptr_1: *const u8 = copy $ptr_2 as *const u8 (PtrToPtr); // _19
            let $addr_1: usize = copy $const_ptr_1 as usize (Transmute); // _20
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            'write:
            (*$ptr_2) = _;
        }
    }
    alloc_cast_check_as_write[$T: type] = {
        fn _(..) -> _ {
            'alloc:
            let $ptr_1: *mut u8 = alloc::alloc::__rust_alloc(_, _); // _3
            let $ptr_2: *mut $T = move $ptr_1 as *mut $T (PtrToPtr); // _2
            let $addr_1: usize = copy $ptr_2 as usize (PointerExposeProvenance); // _6
            // switchInt(move $addr_1) {
            //     0_usize => {}
            //     _ => {}
            // }
            'write:
            (*$ptr_2) = _;
        }
    }
}

diag {
    p = {
        primary(write) = "it is an undefined behavior to dereference a null pointer, and `std::alloc::alloc` may return a null pointer",
        label(alloc)   = "pointer created here",
        label(write)   = "pointer used here",
        note           = "See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc",
        // help           = "check whether it's null before dereferencing",
        name           = "unchecked_allocated_pointer",
        level          = "warn",
    }
}
