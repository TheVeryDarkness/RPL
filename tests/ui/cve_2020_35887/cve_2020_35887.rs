//@ revisions: inline regular
//@[inline] compile-flags: -Z inline-mir=true
//@[regular] compile-flags: -Z inline-mir=false
use std::alloc::{Layout, alloc, alloc_zeroed, dealloc};
use std::ops::{Index, IndexMut, Range};

pub struct Array<T> {
    size: usize,
    ptr: *mut T,
}

impl<T> Array<T> {
    /// Convert to slice
    pub fn to_slice<'a>(&'a self) -> &'a [T] {
        unsafe { std::slice::from_raw_parts(self.ptr as *const T, self.size) }
    }

    /// Convert to mutable slice
    pub fn to_slice_mut<'a>(&'a mut self) -> &'a mut [T] {
        unsafe { std::slice::from_raw_parts_mut(self.ptr, self.size) }
    }

    /// The length of the array (number of elements T)
    pub fn len(&self) -> usize {
        self.size
    }
}

impl<T> Array<T>
where
    T: Default + Copy,
{
    /// Easy initialization if all you want is your T's default instantiation
    // #[rpl::dump_mir(dump_cfg, dump_ddg)]
    pub fn new(size: usize) -> Self {
        let objsize = std::mem::size_of::<T>();
        let layout = Layout::from_size_align(size * objsize, 8).unwrap();
        let ptr = unsafe { alloc(layout) as *mut T };
        //~[regular]^ ERROR: resulting pointer `*mut T` has a different alignment than the original alignment that the pointer was created with
        //~[regular]| ERROR: public function `new` allocates a pointer that may be zero-sized, which is an undefined behavior
        let default: T = Default::default();
        for i in 0..size {
            unsafe {
                (*(ptr.wrapping_offset(i as isize))) = default;
                //~^ ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
                // FIXME: false positive
            }
        }
        Self { size, ptr }
    }
}

impl<T> Array<T>
where
    T: Clone,
{
    /// More generic initialization instantiating all elements as copies of some template
    // #[rpl::dump_mir(dump_cfg, dump_ddg)]
    pub unsafe fn new_from_template(size: usize, template: &T) -> Self {
        let objsize = std::mem::size_of::<T>();
        let layout = Layout::from_size_align(size * objsize, 8).unwrap();
        let ptr = unsafe { alloc(layout) as *mut T };
        for i in 0..size {
            unsafe {
                (*(ptr.wrapping_offset(i as isize))) = template.clone();
                //~[inline]^ ERROR: dropped an possibly-uninitialized value
                // Not a false positive
            }
        }
        Self { size, ptr }
    }
}

impl<T> Index<usize> for Array<T> {
    type Output = T;

    // #[rpl::dump_mir(dump_cfg, dump_ddg)]
    fn index<'a>(&'a self, idx: usize) -> &'a Self::Output {
        unsafe { self.ptr.wrapping_offset(idx as isize).as_ref() }.unwrap()
        //~^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
    }
}

impl<T> IndexMut<usize> for Array<T> {
    fn index_mut<'a>(&'a mut self, idx: usize) -> &'a mut Self::Output {
        unsafe { self.ptr.wrapping_offset(idx as isize).as_mut() }.unwrap()
        //~^ERROR: it is an undefined behavior to offset a pointer using an unchecked integer
    }
}

impl<T> Index<Range<usize>> for Array<T> {
    type Output = [T];

    fn index<'a>(&'a self, idx: Range<usize>) -> &'a Self::Output {
        &self.to_slice()[idx]
    }
}

impl<T> IndexMut<Range<usize>> for Array<T> {
    fn index_mut<'a>(&'a mut self, idx: Range<usize>) -> &'a mut Self::Output {
        &mut self.to_slice_mut()[idx]
    }
}

impl<T> Drop for Array<T> {
    fn drop(&mut self) {
        let objsize = std::mem::size_of::<T>();
        let layout = Layout::from_size_align(self.size * objsize, 8).unwrap();
        unsafe {
            dealloc(self.ptr as *mut u8, layout);
        }
    }
}
