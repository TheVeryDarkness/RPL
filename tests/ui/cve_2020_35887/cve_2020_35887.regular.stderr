error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35887/cve_2020_35887.rs:44:19
   |
LL |         let ptr = unsafe { alloc(layout) as *mut T };
   |                            ----------------------- pointer used here
...
LL |                 (*(ptr.wrapping_offset(i as isize))) = default;
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ offset here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset
   = note: `-D rpl::unchecked-pointer-offset` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::unchecked_pointer_offset)]`

error: resulting pointer `*mut T` has a different alignment than the original alignment that the pointer was created with
  --> tests/ui/cve_2020_35887/cve_2020_35887.rs:38:28
   |
LL |         let ptr = unsafe { alloc(layout) as *mut T };
   |                            -------------^^^^^^^^^^
   |                            |
   |                            pointer created here
   |                            pointer casted here
   |
   = note: See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc
   = note: `-D rpl::misaligned-pointer` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::misaligned_pointer)]`

error: public function `new` allocates a pointer that may be zero-sized, which is an undefined behavior
  --> tests/ui/cve_2020_35887/cve_2020_35887.rs:38:28
   |
LL |     pub fn new(size: usize) -> Self {
   |                ---- size passed in here
...
LL |         let ptr = unsafe { alloc(layout) as *mut T };
   |                            ^^^^^^^^^^^^^ pointer created here
   |
   = note: See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.alloc
   = note: `#[deny(rpl::alloc_maybe_zero)]` on by default

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35887/cve_2020_35887.rs:79:18
   |
LL |         unsafe { self.ptr.wrapping_offset(idx as isize).as_ref() }.unwrap()
   |                  --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  offset here
   |                  pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

error: it is an undefined behavior to offset a pointer using an unchecked integer
  --> tests/ui/cve_2020_35887/cve_2020_35887.rs:86:18
   |
LL |         unsafe { self.ptr.wrapping_offset(idx as isize).as_mut() }.unwrap()
   |                  --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  offset here
   |                  pointer used here
   |
   = help: check whether it's in bound before offsetting
   = note: See the safety section in https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

error: aborting due to 5 previous errors

